drivers: flash: ospi for stm32 mcu

FRASTM opened the pull request:
Fixes command configuration to have samples/drivers/spi_flash
passed on stm32l562 and stm32u585 disco kits.
The code is inspired from the [stm32Cube](https://github.com/STMicroelectronics/STM32CubeL5/tree/master/Drivers/BSP) 


Fixes https://github.com/zephyrproject-rtos/zephyr/issues/46740

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
drivers: flash: ospi for stm32 mcu

Fixes command configuration to have samples/drivers/spi_flash
passed on stm32l562 and stm32u585 disco kits.
In OctoSPI STR/DTR and SPI/STR modes

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
samples/drivers/spi_flash tested on the  stm32l562 and stm32u585 disco kits in OPI/DTR modes

```
*** Booting Zephyr OS build zephyr-v3.1.0-586-gdf7f3121ab7d  ***

JEDEC OSPI-NOR SPI flash testing
==========================

Test 1: Flash erase
Flash erase succeeded!

Test 2: Flash write
Attempting to write 4 bytes                                                                      
Data read matches data written. Good!! 
```

FRASTM commented:
successfully tested on stm32u5  disco kit : b_u585i_iot02
OSPI flash config is OPI / DTR 
OSPI flash config is OPI / STR
OSPI configuring SPI|DUAL|QUAD / STR

tobias-aunbol reviewed (dismissed):
This fixed issues With the AutoPoll fail and flash busy which i also have experienced.
I have tested it on my  B-U585I-IOT02A board with the MX25LM51245 SPI flash

GeorgeCGV commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -276,7 +275,9 @@ static int ospi_read_sfdp(const struct device *dev, off_t addr, uint8_t *data,
 	cmd.Address = addr;
 	cmd.NbData = size;
-
+	cmd.AddressSize = ((dev_cfg->data_mode == OSPI_SPI_MODE)
+				?  HAL_OSPI_ADDRESS_24_BITS
+				:  HAL_OSPI_ADDRESS_32_BITS);
```
Why 32 bits for everything except SPI?
What about dual and quad modes? Many ICs there expect 24 bits.
Possible solution:
```suggestion
	cmd.AddressSize = ((dev_cfg->data_mode == OSPI_OPI_MODE)
				?  HAL_OSPI_ADDRESS_34_BITS
				:  HAL_OSPI_ADDRESS_24_BITS);
```

FRASTM replied:
ok for changing order to 
```
	cmd.AddressSize = ((dev_cfg->data_mode == OSPI_OPI_MODE)
				?  HAL_OSPI_ADDRESS_32_BITS
				:  HAL_OSPI_ADDRESS_24_BITS);
```

GeorgeCGV commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -319,21 +320,18 @@ static int stm32_ospi_mem_ready(OSPI_HandleTypeDef *hospi, uint8_t nor_mode, uin
 	} else {
 		s_command.Instruction = SPI_NOR_CMD_RDSR;
-		s_command.InstructionMode = HAL_OSPI_INSTRUCTION_1_LINE;
-		s_command.DataMode = HAL_OSPI_DATA_1_LINE;
```
Many dual and quad mode ICs use 1 data line mode for that. 
By removing that line a data line mode is enforced to be 2 or 4 lines in the prepare cmd, but shall be 1 line.

FRASTM replied:
Forcing the `s_command.DataMode = HAL_OSPI_DATA_1_LINE` for all non-ospi modes
after the prepare command which sets the  s_command.DataMode

GeorgeCGV replied:
> Do you mean s_command.InstructionMode = HAL_OSPI_INSTRUCTION_1_LINE too ?

Sorry, yes.

GeorgeCGV commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -367,7 +365,6 @@ static int stm32_ospi_write_enable(OSPI_HandleTypeDef *hospi, uint8_t nor_mode,
 		s_command.Instruction = SPI_NOR_OCMD_WREN;
 	} else {
 		s_command.Instruction = SPI_NOR_CMD_WREN;
-		s_command.InstructionMode = HAL_OSPI_INSTRUCTION_1_LINE;
```
Many dual and quad mode ICs use 1 line mode for that.
By removing

> `s_command.InstructionMode = HAL_OSPI_INSTRUCTION_1_LINE;`

the mode is enforced to be 2 or 4 lines in the prepare cmd while it shall be 1 line.

FRASTM replied:
rcing the `s_command.InstructionMode = HAL_OSPI_INSTRUCTION_1_LINE` for all non-ospi modes
after the prepare command which sets the s_command.DataMode

GeorgeCGV commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -378,18 +375,22 @@ static int stm32_ospi_write_enable(OSPI_HandleTypeDef *hospi, uint8_t nor_mode,
 						: SPI_NOR_DUMMY_REG_OCTAL;
 	} else {
 		s_command.Instruction = SPI_NOR_CMD_RDSR;
-		s_command.InstructionMode = HAL_OSPI_INSTRUCTION_1_LINE,
```
same

GeorgeCGV commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -903,9 +903,10 @@ static int flash_stm32_ospi_read(const struct device *dev, off_t addr,
-			cmd.Instruction = dev_data->read_opcode;
-			cmd.DummyCycles = dev_data->read_dummy;
+			/* SPI and STR : use fast read with addr on 4 bytes */
+			cmd.Instruction = SPI_NOR_CMD_READ_FAST_4B;
+			cmd.DummyCycles = SPI_NOR_DUMMY_RD;
```
Why use hardcoded read opcode/instruction here?
Not every flash supports `FAST_4B` (_e.g. some winbond_) and the opcode can be discovered from SFDP...

If we have to hardcode that, then I propose to revert this block back to the SFDP approach but allow to set custom read opcode in the DTS. Otherwise, that change breaks some flash ICs.

FRASTM replied:
Ok restoring the instruction set to dev_data->read_opcode with a comment to remind that SPI_NOR_CMD_READ_FAST_4B is prefered

GeorgeCGV replied:
> Ok restoring the instruction set to dev_data->read_opcode with a comment to remind that SPI_NOR_CMD_READ_FAST_4B is prefered

Ok. I don't think that there is any preferred read opcode as it depends entirely on the flash and/or manufacturer.
Maybe we shall add `readoc` into DTS and prioritize it over SFDP?

FRASTM replied:
Yes, but maybe in _a next PR,_ to fix the initial issue now.
we could add a property in the st,stm32-ospi-nor.yaml like: 
```
    readoc:
      type: int
      required: false
      enum:
        - "READ"  		# 0x03   SPI_NOR_CMD_READ /* Read data */
        - "READ_4B" 		# 0x13   SPI_NOR_CMD_READ_4B  /* Read data 4 Byte Address */
        - "READ_FAST_4B" 	# 0x0C   SPI_NOR_CMD_READ_FAST_4B  /* Fast Read 4 Byte Address */
        - "OCTAL_RD"		# 0xEC13 SPI_NOR_CMD_OCTAL_RD   /* octal Read data */
        - "DREAD" 		# 0x3B SPI_NOR_CMD_DREAD /* Read data (1-1-2) */
        - "2READ" 		# 0xBB SPI_NOR_CMD_2READ /* Read data (1-2-2) */
        - "QREAD" 		# 0x6B SPI_NOR_CMD_QREAD /* Read data (1-1-4) */
        - "4READ" 		# 0xEB SPI_NOR_CMD_4READ /* Read data (1-4-4) */
      description: |
        The value is the instruction to send to the octoflash for read operations.
        If present it will overwrites the value from the SFDP table parameter
```
and then use this property when reading (flash_stm32_ospi_read) instead of value from the SFDP : 
```
#if DT_NODE_HAS_PROP(DT_INST(0, st_stm32_ospi_nor), readoc)
			/* take the instruction from the DTS */
			cmd.Instruction = DT_INST_PROP(0, readoc),
#else
			/* use SFDP:BFP read instruction */
			cmd.Instruction = dev_data->read_opcode;
#endif

```

erwango replied:
@GeorgeCGV I agree with @FRASTM. Rather than trying to fix every potential failure we can meet in current PR, we should focus to fix initial regression (https://github.com/zephyrproject-rtos/zephyr/issues/46740), which is blocking multiple activities (see https://github.com/zephyrproject-rtos/zephyr/issues/48175 for instance).
Then we can deal with other potential cases in a new PR.

GeorgeCGV replied:
@FRASTM, yes `readoc` could be a good workaround for now.
@erwango, sure as there are more OSPI PRs piling up.

GeorgeCGV commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -942,7 +943,11 @@ static int flash_stm32_ospi_write(const struct device *dev, off_t addr,
 	OSPI_RegularCmdTypeDef cmd_pp = ospi_prepare_cmd(dev_cfg->data_mode, dev_cfg->data_rate);
 
-	cmd_pp.Instruction = dev_data->write_opcode;
+	/* using 32bits address also in SPI/STR mode */
+	cmd_pp.Instruction = (dev_cfg->data_mode == OSPI_OPI_MODE)
+			? SPI_NOR_OCMD_PAGE_PRG
+			: SPI_NOR_CMD_PP_4B;
+	/* Do not rely on SFDP:BFP write instruction */
```
why hardcoding write opcode? That is similar to the read opcode situation... That breaks the driver for some ICs.
If SFDP alone is a problem, then the driver shall use the SFDP opcodes but allow users to override that in DTS.

The SFDP is helpful as never before due to semiconductor shortages...

FRASTM replied:
In case of SPi/STR the instruction must be SPI_NOR_CMD_PP_4B, else 
```
Data read does not match data written!!                                         
000ff000 wrote 55 read ff MISMATCH                                              
000ff001 wrote aa read ff MISMATCH                                              
000ff002 wrote 66 read ff MISMATCH                                              
000ff003 wrote 99 read ff MISMATCH   
```

Consequently setting  `cmd_pp.Instruction = dev_data->write_opcode` requires a 
```
case OSPI_SPI_MODE:
			data->write_opcode = SPI_NOR_CMD_PP_4B;
			break;
```

in the `spi_nor_process_bfp` function

GeorgeCGV replied:
> In case of **SPI/STR**, the instruction _must_ be SPI_NOR_CMD_PP_4B, else
> 
> ```
> Data read does not match data written!!                                         
> 000ff000 wrote 55 read ff MISMATCH                                              
> 000ff001 wrote aa read ff MISMATCH                                              
> 000ff002 wrote 66 read ff MISMATCH                                              
> 000ff003 wrote 99 read ff MISMATCH   
> ```
> 
> Consequently setting `cmd_pp.Instruction = dev_data->write_opcode` requires a
> 
> ```
> case OSPI_SPI_MODE:
> 			data->write_opcode = SPI_NOR_CMD_PP_4B;
> 			break;
> ```
> 
> in the `spi_nor_process_bfp` function

@FRASTM, please enable `four-byte-opcodes` in the DTS for your flash, which shall use a 4-byte opcode (`SPI_NOR_CMD_PP_4B`) in the case of `OSPI_SPI_MODE`.

FRASTM replied:
ok done for b_u585i_iot02a and stm32l562e_dk

GeorgeCGV commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -976,8 +981,7 @@ static int flash_stm32_ospi_write(const struct device *dev, off_t addr,
 	}
 
-	cmd_pp.Address = addr;
-	cmd_pp.AddressSize = stm32_ospi_hal_address_size(dev);
```
Not every flash IC works with a 32-bit address size. Not sure why the setting of the size is removed here...

FRASTM replied:
In case of OPI/STR transfer, the cmd.AddressSize must be HAL_OSPI_ADDRESS_32_BITS 
and stm32_ospi_hal_address_size gives HAL_OSPI_ADDRESS_24_BITS  which fails when reading:
```
Data read does not match data written!!                                         
000ff000 wrote 55 read ff MISMATCH                                              
000ff001 wrote aa read ff MISMATCH                                              
000ff002 wrote 66 read ff MISMATCH                                              
000ff003 wrote 99 read ff MISMATCH 
```

GeorgeCGV replied:
I don't know what Flash you use, but MX25LM51245G does support 24 and 32 bits address sizes depending on the configuration. Winbond's W25Q128JV supports only 24 bits address size.

Therefore, it is not so much Mode dependent rather than Flash IC-specific.

Considering some of your previous comments your flash works with 4-byte opcodes and that requires 4-byte address width (_otherwise it is a configuration error_).

Currently, the address width is taken from SFDP,BFP:

> `spi_nor_process_bfp_addrbytes(dev, jesd216_bfp_addrbytes(bfp));`

Please check your SFDP table and enable `four-byte-opcodes` in the DTS.

The alternative is to allow to set address width in the DTS and prioritize it over SFDP, BFP.

GeorgeCGV replied:
@FRASTM, ping.

Alternative: 

Quad and dual flashes might not work with a 32-bits address size.
ospi_prepare_cmd enforces 32-bits address size. Please use 24-bits for non OSPI_OPI_MODE.

FRASTM replied:
ok

GeorgeCGV commented:
Unfortunately, I don't have any board with OSPI flash IC. Can test only against quad/dual ICs over OSPI (_i.e. MX25L12835F_).
However, current modifications will break OSPI for quad and dual modes, especially for some Winbond ICs (_lack of support for the hardcoded opcodes_).

FRASTM commented:
set the  `s_command.InstructionMode = HAL_OSPI_INSTRUCTION_1_LINE` for all non-ospi modes

carlescufi commented:
@erwango ping

GeorgeCGV commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -319,21 +320,22 @@ static int stm32_ospi_mem_ready(OSPI_HandleTypeDef *hospi, uint8_t nor_mode, uin
 	/* Configure automatic polling mode command to wait for memory ready */
 	if (nor_mode == OSPI_OPI_MODE) {
 		s_command.Instruction = SPI_NOR_OCMD_RDSR;
+		s_command.AddressMode = HAL_OSPI_ADDRESS_8_LINES;
```
`ospi_prepare_cmd` already sets `s_command.AddressMode = HAL_OSPI_ADDRESS_8_LINES` for `OSPI_OPI_MODE`.

FRASTM replied:
ok

GeorgeCGV commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -378,18 +381,26 @@ static int stm32_ospi_write_enable(OSPI_HandleTypeDef *hospi, uint8_t nor_mode,
+	/* New command to Configure automatic polling mode to wait for write enabling */
 	if (nor_mode == OSPI_OPI_MODE) {
 		s_command.Instruction = SPI_NOR_OCMD_RDSR;
+		s_command.AddressMode = HAL_OSPI_ADDRESS_8_LINES;
+		s_command.DataMode = HAL_OSPI_DATA_8_LINES;
```
`ospi_prepare_cmd` already sets `s_command.AddressMode = HAL_OSPI_ADDRESS_8_LINES` and `s_command.DataMode = HAL_OSPI_DATA_8_LINES`  for `OSPI_OPI_MODE`.

FRASTM replied:
yes for the WREN command but here is the RDSR command to be prepared,

GeorgeCGV commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -887,8 +898,7 @@ static int flash_stm32_ospi_read(const struct device *dev, off_t addr,
 
 	/* Instruction and DummyCycles are set below */
-	cmd.Address = addr;
-	cmd.AddressSize = stm32_ospi_hal_address_size(dev);
```
Quad and dual flashes might not work with a 32-bits address size.
`ospi_prepare_cmd` enforces 32-bits address size. Please use 24-bits for non `OSPI_OPI_MODE`.

FRASTM replied:
ok

GeorgeCGV commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -1467,6 +1479,9 @@ static int spi_nor_process_bfp(const struct device *dev,
 		case OSPI_DUAL_MODE:
 			data->write_opcode = SPI_NOR_CMD_PP_1_1_2;
 			break;
+		case OSPI_SPI_MODE:
+			data->write_opcode = SPI_NOR_CMD_PP_4B;
+			break;
```
Shall be removed.

It is `SPI_NOR_CMD_PP_4B` only when 4-byte address width is used.
The driver will automatically convert `SPI_NOR_CMD_PP` to `SPI_NOR_CMD_PP_4B` when `four-byte-opcodes` is set in the DTS (see line `1522`).

FRASTM replied:
ok

FRASTM commented:
rebased + change after review

FRASTM committed:
boards: arm: stm32u585 and stm32l562 disco octoflash in SPI mode

The four-byte-opcodes property of the octoFlash node must be set to
use a 4-byte opcode (SPI_NOR_CMD_PP_4B) in the case of OSPI_SPI_MODE.
Else the write instruction is wrong and the operation fails
in this transfer data mode

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
successfully tested on stm32u5 disco kit : b_u585i_iot02 and stm32l562e_dk
OSPI flash config is OPI / DTR
OSPI flash config is OPI / STR
OSPI configuring SPI|DUAL|QUAD / STR
```
*** Booting Zephyr OS build zephyr-v3.1.0-3175-g923410906b43  ***               
                                                                                
ospi-nor-flash@0 SPI flash testing                                              
==========================                                                      
                                                                                
Test 1: Flash erase                                                             
Flash erase succeeded!                                                          
                                                                                
Test 2: Flash write                                                             
Attempting to write 4 bytes                                                     
Data read matches data written. Good!! 
```
